PRL: Document noise function eligibility warnings (#1289)
ihmeuw/vivarium_research: docs/source/models/concept_models/vivarium_census_synthdata/concept_model.rst

<census_prl_guardian_duplication>`. Because of our distributed data processing
(the data is split into multiple shards), it is not entirely straightforward to
determine the overall fraction of rows that are eligible for the requested noise
or to add noise to the correct number of rows across all shards. Here is a
simple strategy to deal with this issue:

#.  Pre-compute the fraction :math:`F` of rows eligible for each applicable
    noise type in the full dataset (i.e., across all shards concatenated
    together).

#.  Check whether :math:`p\le F`, where :math:`p` is the average fraction of
    rows the user wants noised (in pseudopeople, :math:`p` is called
    ``cell_probability`` for column-based noise or ``row_probability`` for
    row-based noise). If :math:`p>F`, raise a warning to the user before
    noising starts.

#.  When noising each shard, compute the fraction :math:`f` of rows eligible
    for noise in the shard, and apply the noise to each eligible row in the
    shard with probability :math:`\min\{p/f, 1\}`. Do **not** raise a warning to
    the user for every shard with :math:`p>f`.

This strategy has a known limitation: It systematically adds less noise to the

> dataset than requested when :math:`p` sufficiently large. This is obvious when

Reviewer: ```suggestion
dataset than requested when :math:`p` is sufficiently large. This is obvious when
```

Due to random fluctuation between the shards, sometimes we will have
:math:`f<F`. Thus, if :math:`p` is sufficiently large, we will have :math:`p>f`
for some shards, even if :math:`p\le F`. Now note that for each shard with
:math:`p>f`, the fraction of rows that get noised is :math:`f`, which is less
than :math:`p`. On the other hand, for shards with :math:`p\le f`, the average
fraction of rows that get noised is :math:`p`. Combining all shards, the average
fraction of noised rows will be less than :math:`p`. This problem will be worse
for small datasets like ACS because of greater variability of :math:`f` around
:math:`F`.

We are willing to accept the above limitation for simplicity's sake. Moreover,
the impact of this issue should be mitigated when we change our shard-based
approach to data storage so that the different data files have similar sizes
rather than some datasets having tiny shards like ACS currently does.

Another potentially ambiguous point in the above strategy is which data to use
to pre-compute the eligible fraction :math:`F`. Since these pre-computed values
are merely for user warnings, the researchers are flexible as to how closely the

> :math:`F` values should match match the corresponding datasets, and we'll let

Reviewer: ```suggestion
:math:`F` values should match the corresponding datasets, and we'll let
```

the engineers decide on the exact implementation. Here are a couple options: If
the pre-computed values are stored in a separate metadata file generated with
each simulation run, then :math:`F` can be computed for each applicable
(dataset, year, noise type) combination that can be requested by the user. On
the other hand, if it's easier to store a single value of :math:`F` for each
(dataset, noise type) without varying across years, then we should should use
the dataset for the year 2030; since 2030 is the midpoint of our simulation, we
expect it to be sufficiently representative of the data across all simulated

> years. This level of imprecision is sufficient for the user warning, which

Reviewer: ```suggestion
years.
If even this is too complicated, a single :math:`F` can be computed for a noise type, and not vary between datasets;
the decennial census dataset should be used to do this.
The level of imprecision is acceptable for the user warning, which
```

Isn't this what was actually implemented for nicknames? I'm pretty sure they used just the 2030 decennial.

Author: Oh yeah, I think you're right -- I forgot the details of what was actually implemented.
